crates/revm/revm-inspectors/src/tracing/builder/geth.rs

```diff
@@ -106,4 +121,60 @@ impl GethTraceBuilder {
             struct_logs,
         }
     }
+
+    pub fn geth_call_traces(
+        &self,
+        receipt_gas_used: U256,
+        opts: CallConfig,
+    ) -> CallFrame {
+        if self.nodes.is_empty() {
+            return Default::default()
+        }
+
+        // first fill up the root
+        let main_trace_node = &self.nodes[0];
+        let main_trace = &main_trace_node.trace;
+
+        let mut root_call_frame = CallFrame {
+            typ: main_trace.kind.to_string(),
+            from: main_trace.caller,
+            to: Some(main_trace.address),
+            value: Some(main_trace.value),
+            gas: U256::from(receipt_gas_used),
+            gas_used: U256::from(main_trace.gas_used),
+            input: Bytes::from(main_trace.data.clone()),
+            output: Some(Bytes::from(main_trace.output.clone())),
+            error: None,
+            revert_reason: None,
+            calls: None,
+            logs: None,
+        };
+
+        // we need to populate error and revert reason
+        if !main_trace.success {
+            root_call_frame.revert_reason = decode_revert_reason(main_trace.output.clone());
+            root_call_frame.error = main_trace.as_error();
+        }
+
+        // logs
+        if opts.with_log.unwrap_or_default() {
+            let mut call_logs = vec![];
+            for log in &main_trace_node.logs {
+                call_logs.push(CallLogFrame {
+                    address: Some(main_trace.address.into()),
+                    topics: Some(log.topics.clone()),
+                    data: Some(Bytes::from(log.data.clone()))
+                });
+            }
+            root_call_frame.logs = Some(call_logs);
+        }
+
+        if opts.only_top_call.unwrap_or_default() {
+            return root_call_frame;
+        }
+
+        // TODO - recursively build the call tracker frames
```

Comment:
please try to write this iteratively, see doc on `add_to_geth_trace`

Reply:
yes will do that!